sources/packages/backend/apps/api/src/institution-integration/ier-integration/ier12-file.service.ts

```diff
@@ -0,0 +1,171 @@
+import { Injectable } from "@nestjs/common";
+import { StudentAssessment } from "@sims/sims-db";
+import { InjectLogger } from "../../common";
+import { LoggerService } from "../../logger/logger.service";
+import { ConfigService, StudentAssessmentService } from "../../services";
+import { InstitutionIntegrationConfig } from "../../types";
+import { getFileNameAsCurrentTimestamp } from "../../utilities";
+import { IER12IntegrationService } from "./ier12-integration.service";
+import {
+  IER12Record,
+  IER12UploadResult,
+} from "./models/ier12-integration.model";
+
+@Injectable()
+export class IER12FileService {
+  institutionIntegrationConfig: InstitutionIntegrationConfig;
+  constructor(
+    config: ConfigService,
+    private readonly ierIntegrationService: IER12IntegrationService,
+    private readonly studentAssessmentService: StudentAssessmentService,
+  ) {
+    this.institutionIntegrationConfig =
+      config.getConfig().InstitutionIntegrationConfig;
+  }
+
+  /**
+   * 1. Fetches the assessment data for the institution location.
+   * 2. Create the Request content for the IER 12 file by populating the
+   * header, footer and trailer content.
+   * 3. Create the request filename with the file path with respect to the institution code
+   * for the IER 12 sent File.
+   * 4. Upload the content to the zoneB SFTP server.
+   * @param generatedDate date in which the assessment for
+   * particular institution is generated.
+   * @returns Processing IER 12 result.
+   */
+  async processIER12File(generatedDate?: Date): Promise<IER12UploadResult[]> {
+    this.logger.log(`Retrieving pending assessment for IER 12...`);
+    const pendingAssessments =
+      await this.studentAssessmentService.getPendingAssessment(generatedDate);
+    if (!pendingAssessments.length) {
+      return [
+        {
+          generatedFile: "none",
+          uploadedRecords: 0,
+        },
+      ];
+    }
+    this.logger.log(`Found ${pendingAssessments.length} assessments.`);
+    const fileRecords: Record<string, IER12Record[]> = {};
+    pendingAssessments.forEach((pendingAssessment) => {
+      const institutionCode =
+        pendingAssessment.offering.institutionLocation.institutionCode;
+      if (!fileRecords[institutionCode]) {
+        fileRecords[institutionCode] = [];
+      }
+      fileRecords[institutionCode].push(
+        this.createIER12Record(pendingAssessment),
+      );
+    });
+    const uploadResult: IER12UploadResult[] = [];
+    try {
+      this.logger.log("Creating IER 12 content...");
+      const fileUploadedContent = Object.keys(fileRecords).map(
+        (institutionCode) =>
+          this.uploadIER12Content(institutionCode, fileRecords),
+      );
+      for await (const ierUploadResult of fileUploadedContent) {
```

Comment:
as a 2nd thought, I was thinking can't we do something like this, 
![image](https://user-images.githubusercontent.com/77353155/201754062-a85f1d9d-c982-4575-94e2-46efbdf3dec0.png)
. and we can avoid 2 loops